File: engine/src/model/symbol.rs
use std::fmt;
use crate::model::{ModuleId, RawRelation, Relation, RelationId};

use super::run::RunId;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct SymbolId(String);

impl SymbolId {
    pub fn new(module_id: &ModuleId, kind: &SymbolKind, name: &str, start_line: usize) -> Self {
        Self(format!(
            "{}::{}::{}::{}",
            module_id,
            kind,
            name,
            start_line
        ))
    }
}

impl fmt::Display for SymbolId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.0)
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SymbolKind {
    Class,
    Function,
    Method,
}

impl fmt::Display for SymbolKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            SymbolKind::Class => f.write_str("class"),
            SymbolKind::Function => f.write_str("function"),
            SymbolKind::Method => f.write_str("method"),
        }
    }
}

#[derive(Debug, Clone)]
pub struct Symbol {
    pub id: SymbolId,
    pub run_id: RunId,
    pub module_id: ModuleId,
    pub parent_symbol_id: Option<SymbolId>,
    pub name: String,
    pub kind: SymbolKind,
    pub doc: String,
    pub location: String,
    pub start_line: usize,
    pub end_line: usize,
}

impl Symbol {
    pub fn new(
        module_id: &ModuleId,
        run_id: &RunId,
        parent_symbol_id: Option<&SymbolId>,
        kind: SymbolKind,
        name: String,
        doc: String,
        location: String,
        start_line: usize,
        end_line: usize,
    ) -> Self {
        let id = SymbolId::new(&module_id, &kind, &name, start_line);
        Self {
            id,
            run_id: run_id.clone(),
            module_id: module_id.clone(),
            parent_symbol_id: parent_symbol_id.cloned(),
            name,
            kind,
            doc,
            location,
            start_line,
            end_line,
        }
    }
}


#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct RawSymbolId(String);

impl RawSymbolId {
    pub fn new(kind: &SymbolKind, name: &str, start_line: usize) -> Self {
        Self(format!(
            "{}::{}::{}",
            kind,
            name,
            start_line
        ))
    }
}

impl fmt::Display for RawSymbolId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.0)
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct RawSymbol {
    pub id: RawSymbolId,
    pub name: String,
    pub kind: SymbolKind,
    pub doc: String,
    pub location: String,
    pub children_symbols: Vec<RawSymbol>,
    pub children_relations: Vec<RawRelation>,
    pub start_line: usize,
    pub end_line: usize,
}

impl RawSymbol {
    pub fn new(
        name: String,
        kind: SymbolKind,
        doc: String,
        location: String,
        start_line: usize,
        end_line: usize,
    ) -> Self {
        let id = RawSymbolId::new(&kind, &name, start_line);
        Self {
            id,
            name,
            kind,
            doc,
            location,
            children_symbols: Vec::new(),
            children_relations: Vec::new(),
            start_line,
            end_line,
        }
    }

    pub fn add_children_symbol(&mut self, child: RawSymbol) {
        self.children_symbols.push(child);
    }

    pub fn add_children_relation(&mut self, relation: RawRelation) {
        self.children_relations.push(relation);
    }

    pub fn into_symbol(
        self,
        module_id: &ModuleId,
        run_id: &RunId,
        parent_symbol_id: Option<&SymbolId>,
    ) -> (Symbol, Vec<RawSymbol>, Vec<RawRelation>) {

        let symbol = Symbol::new(
            module_id,
            run_id,
            parent_symbol_id,
            self.kind,
            self.name,
            self.doc,
            self.location,
            self.start_line,
            self.end_line,
        );
        (symbol, self.children_symbols, self.children_relations)
    }
}

File: api/src/api/models/symbol.py
from dataclasses import dataclass
from typing import Optional, Dict, Any

from .run import RunId
from enum import Enum
from .module import ModuleId

class SymbolKind(Enum): 
    Class = "class"
    Function = "function"
    Method = "method"



class SymbolId(str):
    "module_id: &ModuleId, kind: &SymbolKind, name: &str, start_line: usize"

    def __new__(cls, module_id: ModuleId, kind: SymbolKind, name: str, start_line: int):
        return super().__new__(cls, f"{module_id}:{kind}:{name}:{start_line}")
    
    @classmethod
    def from_db_row(cls, symbol_id: str) -> "SymbolId":
        module_id, kind, name, start_line = symbol_id.rsplit(":", 3)
        return cls(ModuleId.from_db_row(module_id), SymbolKind(kind), name, int(start_line))

@dataclass
class Symbol:
    id: SymbolId
    run_id: RunId
    module_id: ModuleId
    parent_symbol_id: Optional[SymbolId]
    name: str
    kind: SymbolKind
    doc: str
    location: str
    start_line: int
    end_line: int


    @classmethod
    def create(cls,run_id: RunId, row: Dict[str, Any]) -> "Symbol":
        return cls(
            id=SymbolId.from_db_row(symbol_id=row['id']),
            run_id=run_id,
            module_id=ModuleId.from_db_row(module_id=row['module_id']),
            parent_symbol_id=SymbolId.from_db_row(symbol_id=row['parent_symbol_id']) if row['parent_symbol_id'] else None,
            name=row['name'],
            kind=SymbolKind(row['kind']),
            doc=row['doc'],
            location=row['location'],
            start_line=row['start_line'],
            end_line=row['end_line']
        )